### PR TITLE
Fix datamodules ImportError and add tests

### LINKED ISSUE #379: Cannot import name 'SklearnDataModule' from 'pl_bolts.datamodules'
## 🐛 Bug

I cannot import a key module, even after a fresh install.

### To Reproduce

Install package to fresh virtualenv and try import. Import fails. See code below.

This is on macOS Catalina Terminal with python 3.8.5

#### Code sample

```shell
$ python3 -m venv venv  
$ source venv/bin/activate 
(venv) $ pip install pytorch-lightning-bolts 
(venv) $ python
Python 3.8.5 (default, Jul 21 2020, 10:48:26) 
[Clang 11.0.3 (clang-1103.0.32.62)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
 >>> from pl_bolts.datamodules import SklearnDataModule
```

Error 1: This fails with:

```
>>> from pl_bolts.datamodules import SklearnDataModule


/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/models/mnist_module.py:14: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/callbacks/variational.py:9: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/callbacks/vision/image_generation.py:9: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/datamodules/binary_mnist_datamodule.py:12: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/datamodules/cifar10_dataset.py:13: UserWarning: You want to use `Pillow` which is not installed yet, install it with `pip install Pillow`.
  warn('You want to use `Pillow` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/transforms/dataset_normalizations.py:5: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/datamodules/cifar10_datamodule.py:17: UserWarning: You want to use `torchvision` which is not installed yet, install it with `pip install torchvision`.
  warn('You want to use `torchvision` which is not installed yet,'  # pragma: no-cover
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'SklearnDataModule' from 'pl_bolts.datamodules' (/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/datamodules/__init__.py)
```

So I install `torchvision`  (is it strictly necessary just for the SklearnModule?)

```shell
(venv) $ pip install torchvision 
(venv) $ python                 
>>> from pl_bolts.datamodules import SklearnDataModule
```

Error 2: This fails with:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'SklearnDataModule' from 'pl_bolts.datamodules' (/Users/paczuskp/projects_gene/raan-lung-2020/code/py/venv/lib/python3.8/site-packages/pl_bolts/datamodules/__init__.py)
```

### Expected behavior

Import loads with no error

### Environment

```
(venv) $ pip freeze
absl-py==0.11.0
cachetools==4.1.1
certifi==2020.11.8
chardet==3.0.4
dataclasses==0.6
fsspec==0.8.4
future==0.18.2
google-auth==1.23.0
google-auth-oauthlib==0.4.2
grpcio==1.33.2
idna==2.10
Markdown==3.3.3
numpy==1.19.4
oauthlib==3.1.0
Pillow==8.0.1
protobuf==3.14.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pytorch-lightning==1.0.7
pytorch-lightning-bolts==0.2.5
PyYAML==5.3.1
requests==2.25.0
requests-oauthlib==1.3.0
rsa==4.6
six==1.15.0
tensorboard==2.4.0
tensorboard-plugin-wit==1.7.0
torch==1.7.0
torchvision==0.8.1
tqdm==4.52.0
typing-extensions==3.7.4.3
urllib3==1.26.2
Werkzeug==1.0.1
```

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #379.

Bacause raising `ModuleNotFoundError` in `pl_bolts/datamodules/sklearn_datamodule.py` makes `from pl_bolts.datamodules import SklearnDataModule` raise `ImportError` as reported in the issue, this PR fixes the problem by replacing it with `warn_missing_pkg`.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by ydcjeff — APPROVED** [review]

**Comment by Borda** [comment]
shall we add some tests that we do not break these imports in the future?

**Comment by akihironitta** [comment]
@Borda Yes, let's.

**Comment by Borda** [comment]
> @Borda Yes, let's.

may we add it as part of this PR?
probably for this test, we would need to mock some packages aka they are not installed...
there shall be a way wrap such test and inside call these testing imports

**Comment by akihironitta** [comment]
@Borda Just added the import tests when optional packages are unavailable. How does it look?
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/ff8cd589328a32557c9f1f68356f94a296f07ab4/tests/datamodules/test_imports.py

**Comment by akihironitta** [comment]
We still have similar problems in importing datamodules when dependencies are not met... e.g. `from pl_bolts.datamodules import DiscountedExperienceSource` still fails without `gym` installed, so let me extend this PR to fix the problem for all datamodules.

**Comment by Borda** [comment]
> @Borda Just added the import tests when optional packages are unavailable. How does it look?
> https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/ff8cd589328a32557c9f1f68356f94a296f07ab4/tests/datamodules/test_imports.py

I may be thinking about some automatic way lets say parse the `__init__` but that would require more work of finding dependency in the follow-up file... co this is fine :]
cc: @oke-aditya

**Comment by oke-aditya** [comment]
I too will give a thought. It doesn't seem very straightforward to me 😅

**Comment by akihironitta** [comment]
I'm aware that there is inconsistency in how we check if optional packages are available (i.e. `try: ... except ModuleNotFoundError: ...` and `importlib.util.find_spec("some_package")`). I'm trying to refactor datamodules in #338. Also, in #346, I'm suggesting to implement a context manager class which will reduce repetitive import checks all over the codebase, so for this PR, let's only make datamodules importing work.

File: pl_bolts/datamodules/kitti_datamodule.py [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
 import torch
-import torchvision.transforms as transforms
 from pytorch_lightning import LightningDataModule
 from torch.utils.data import DataLoader
 from torch.utils.data.dataset import random_split
 
 from pl_bolts.datasets.kitti_dataset import KittiDataset
+from pl_bolts.utils.warnings import warn_missing_pkg
+
+_TORCHVISION_AVAILABLE = importlib.util.find_spec("torchvision") is not None
```
Comment (Borda) [inline pl_bolts/datamodules/kitti_datamodule.py:12]: let's move these available packages to `pl_bolts/__init__`

File: pl_bolts/datamodules/sklearn_datamodule.py [hunk 2]
```diff
@@ -6,11 +6,12 @@
 from pytorch_lightning import LightningDataModule
 from torch.utils.data import DataLoader, Dataset
 
+from pl_bolts.utils.warnings import warn_missing_pkg
+
 try:
     from sklearn.utils import shuffle as sk_shuffle
 except ModuleNotFoundError:
```
Comment (Borda) [inline pl_bolts/datamodules/sklearn_datamodule.py:13]:
```suggestion
_SKLEARN_AVAILABLE = importlib.util.find_spec("sklearn") is not None
if not _SKLEARN_AVAILABLE:
    from sklearn.utils import shuffle as sk_shuffle
else:
```
lest add this replacement everywhere, in another PR :]

File: pl_bolts/datamodules/sklearn_datamodule.py [hunk 3]
```diff
@@ -6,11 +6,12 @@
...
-    raise ModuleNotFoundError('You want to use `sklearn` which is not installed yet,'  # pragma: no-cover
-                              ' install it with `pip install sklearn`.')
+    warn_missing_pkg("sklearn")
```
Comment (Borda) [inline pl_bolts/datamodules/sklearn_datamodule.py:14]:
```suggestion
    warn_missing_pkg("sklearn")  # pragma: no-cover
```

File: tests/datamodules/test_imports.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    ("TensorDataset", []),
+    ("SSLImagenetDataModule", ["torchvision"]),
+    ("STL10DataModule", ["torchvision"]),
+    ("VOCDetectionDataModule", ["torchvision"]),
+    ("CityscapesDataModule", ["torchvision"]),
+    ("KittiDataset", ["PIL"]),
+    ("KittiDataModule", ["torchvision"]),
+])
+def test_import(dm_cls, deps):
+    with mock.patch.dict("sys.modules", {pkg: None for pkg in deps}):
```
Comment (Borda) [inline tests/datamodules/test_imports.py:29]: pls, add here doc/sentence explaining what this test for...

File: pl_bolts/datamodules/sklearn_datamodule.py (hunk 2 above)
Comment (akihironitta) [reply pl_bolts/datamodules/sklearn_datamodule.py:13]: Sure, let's do this in the ongoing PR #338.

File: pl_bolts/datamodules/kitti_datamodule.py (hunk 1 above)
Comment (akihironitta) [reply pl_bolts/datamodules/kitti_datamodule.py:12]: Shall we do this here or in #338?

Comment (akihironitta) [reply pl_bolts/datamodules/kitti_datamodule.py:12]: @Borda I'll quickly do this in #338 for all optional packages!

**Comment by akihironitta** [comment]
@Borda mind having a look again?

**Pull request merged**